apply: Support overriding interface configuration without merging

cathay4t opened the pull request:
When using `nmstatectl gc`, unmentioned IP section means disabled, but
in `nmstatectl service` or `nmstatectl apply`, unmentioned IP means
merging from current. In order to be consistent when migrating from
`nmstatectl gc` to `nmstatectl service`, we need to introduce a option
to skip merging IP.

We also have user request us to not merging existing route by only
apply route found in desired state.

So we introduce option to skip merging current state:

 * Rust: `NetworkState.set_override_iface(true)`
 * Python: `libnmstate.apply(state, override_iface=True)`
 * C: `nmstate_net_state_apply()` with `NMSTATE_FLAG_OVERRIDE_IFACE` flag.
 * CLI: `nmstatectl apply --override-iface <state_file>`
 * nmstate.service: `/etc/nmstate/nmstate.conf` with:

    ```toml
    [service]
    override_iface = true
    ```

When in `override_iface` mode, nmstate will:

 * For `interfaces` section, any defined interface will be configured by
   desire state information only.
 * Other sections will ignore this `override_iface` mode and still
   merging as it was.

Technical detail:
 * When `override_iface: true`, we purge current network to only keep
   interface name and type before merging it with desire state.
 * When `override_iface: true`, we do not re-use existing NetworkManager
   connection.

Integration test cases included.

Resolves: https://issues.redhat.com/browse/RHEL-59935

ihuguet commented on rust/src/lib/gen_conf.rs:
```diff
@@ -29,8 +29,7 @@ impl NetworkState {
             cur_state,
-            true,  // gen_conf mode
-            false, // memory only
+            true, // gen_conf mode
```
Why is this needed for?

cathay4t replied:
Removed.

ihuguet commented on rust/src/lib/net_state.rs:
```diff
@@ -212,6 +214,14 @@ impl NetworkState {
         self
     }
 
+    /// When set to true, the network state be applied will not merging current
```
The sentence is not well constructed and it's difficult to understand. Can you reword it?

cathay4t replied:
Removed. New approach been used.

ihuguet commented on rust/src/lib/ip.rs:
```diff
@@ -200,6 +200,40 @@ impl InterfaceIpv4 {
 
+    /// Set explicit default values if not desired. Used when do not want to
+    /// merge ip settings from current status
+    pub fn set_explicit_defaults(&mut self) {
```
I think that the name should be `set_implicit_defaults`, actually, isn't it? Some defaults are set implicitly depending on other values, they are not explicitly specified...

cathay4t replied:
Removed. New approach been used.

ihuguet commented on rust/src/lib/ip.rs:
```diff
@@ -200,6 +200,40 @@ impl InterfaceIpv4 {
+            self.enabled = true
+        }
+
+        if self.dhcp.is_none() {
+            self.dhcp = Some(false)
+        }
```
Does it make sense to enable dhcp if `enabled=false`?

cathay4t replied:
Removed. New approach been used.

ihuguet commented on rust/src/lib/iface.rs:
```diff
@@ -781,7 +781,24 @@ impl MergedInterface {
+        override_ip: bool,
     ) -> Result<Self, NmstateError> {
+        let mut desired = desired.clone();
+        if override_ip {
+            if let Some(state) = desired.as_mut() {
+                state
+                    .base_iface_mut()
+                    .ipv4
+                    .get_or_insert(Default::default())
+                    .set_explicit_defaults();
+                state
+                    .base_iface_mut()
+                    .ipv6
+                    .get_or_insert(Default::default())
+                    .set_explicit_defaults();
+            }
+        }
```
I don't have time to test it now, but it seems to me that we are not actually "overriding". We are still merging with current network state. The actual difference is that we asume some implicit default values in the desired state [1].

Thus, "override" might be a bit missleading. Maybe something like `--ip-implicit-defaults`?

Also, if this option only changes the desired state, we don't need to do this here, we can do it wherever the desired state is created. This would also avoid having to pass the variable `override_ip` over and over.

[1] OK, one of the implicit default value is `enabled: false`, which basically creates the impression of overriding the IP config, but still I don't find it a precise name.

cathay4t replied:
Removed. New approach been used. Please check again.

ihuguet commented on rust/src/cli/apply.rs:
```diff
@@ -60,52 +62,34 @@ where
-    reader.read_to_string(&mut content)?;
-    let content = content.replace('\u{A0}', " ");
-
-    let mut net_state: NetworkState = match serde_yaml::from_str(&content) {
```
There are 37 changed files, this is hard to review properly. Can you please split into several commits so they can be reviewed in smaller logical steps? Also, consider that a single PR might need to be reviewed more than once, so it gets even harder.

Some of the changes like this one seems to be a refactor, and also not really related to the goal of the PR. I'm fine including such things into the PR, but better in a separate dedicated commit.

cathay4t replied:
New approach only has 16 files now. I will isolate out the CLI unrelated changes to new patch.

cathay4t commented:
We got https://issues.redhat.com/browse/RHEL-57565 requesting similar on `no merge` for route rule. Let me change the API design for this PR to fit that usage also.

tyll commented on rust/src/cli/ncl.rs:
```diff
@@ -193,6 +193,14 @@ fn main() {
+                        .long("skip-merge")
+                        .takes_value(false)
+                        .help(
+                            "Do not merge current network state"
```
This help text is not sufficient to understand what this option does.

cathay4t replied:
This is a brief message, I have added manpage with detail for this.

tyll commented:
I don't understand what skip merge is supposed to mean from the description here. The technical details describe some effects of it. What's lacking is a description that defines the concept and when and when not to use it. To me it seems a little bit that it is supposed to indicate that the applied state is considered to be full for the specified interfaces but I don't understand what it means for the routes. Especially, what is the effect of `nmstatectl apply --kernel` vs `nmstatectl apply --kernel --skip-merge`.

cathay4t commented:
> I don't understand what skip merge is supposed to mean from the description here. The technical details describe some effects of it. What's lacking is a description that defines the concept and when and when not to use it. To me it seems a little bit that it is supposed to indicate that the applied state is considered to be full for the specified interfaces but I don't understand what it means for the routes. Especially, what is the effect of `nmstatectl apply --kernel` vs `nmstatectl apply --kernel --skip-merge`.

Changed the commit comment to:

```
If skip-merge set true, nmstate will not merge current state when applying desired state:

 * For `interfaces` section, any defined interface will be configured by desire
   state information only.
 * For `routes` section, all routes of impacted interface(next-hop-to,
   otherwise loopback interface) will be override by desired state.
 * For `route-rules` section, all route rules of impacted interface will be
   override by desired state. The impacted interface is loopback unless desire
   state has matching interface and route sections holding with route table
   used by route rule.
 * For other sections, if defined in desire state, nmstate just override that
   section configure.
```

There is no special handling on kernel mode, it will behave as state above.

cathay4t commented:
@tyll Will it easier for you if I change `--skip-merge` to `--override`?

ihuguet commented on rust/src/lib/net_state.rs:
```diff
@@ -212,6 +214,18 @@ impl NetworkState {
 
+    /// Nmstate by default will merge information including `ipv4`,
+    /// `ipv6`, `routes`, `route-rules` and `ovn` sections.
+    /// When set to true, the network state be applied without merging current
```
typo: `be` -> `will be`

ihuguet commented on rust/src/lib/net_state.rs:
```diff
@@ -212,6 +214,18 @@ impl NetworkState {
+    /// When set to true, the network state be applied without merging current
+    /// network state. For example, if desired state has no IP section defined,
+    /// nmstate will treat it as disabled.
+    /// When set to true, the `state: absent` for routes and route rules will
+    /// not be take effect.
```
will not ~be~ take effect.

ihuguet replied:
Why `state: absent` won't take effect? As I understand what this is doing, it will remove routes and rules for the interfaces mentioned in the desired state. But if routes for other interfaces are mentioned as `absent`, why not remove them?

cathay4t replied:
The `state: absent` is partial editing, it should not exist in overriding mode.

Meanwhile, the `state: absent` could modify interface not mentioned in desired state which might not expected by user. So for now, I think intend to leave it as not supported and wait user request on this.

ihuguet reviewed (changes requested):
I see some changes related to rules, but nothing related to routes. Is not anything that we need to consider?

Also, I think there are very few tests. To cover everything, tens of tests would be needed. That's too mucho to add them now, but some more cases involving different configurations, different cases regarding routes and rules, would be good.

cathay4t commented:
> I see some changes related to rules, but nothing related to routes. Is not anything that we need to consider?

The test case `test_override_route` already confirmed the route work as expected.

The reason we do not need code changes to route is because nmstate top level code is handling the merge and absent action as it is very easy to tell which interface should be used for this route.

For route rule, many of them does not have interface related information, the processing is handled at nm code level.

> 
> Also, I think there are very few tests. To cover everything, tens of tests would be needed. That's too mucho to add them now, but some more cases involving different configurations, different cases regarding routes and rules, would be good.

I only added test case for requested use case. I would like wait user bug report to add more test cases for this mode.
Instead of guessing use case, I prefer fix only with bug report.

cathay4t committed:
cli: Unify the state reading function

Unify the state reading code in `apply` and `service` command to shared
function `state_from_fd()`.

Existing integration test case are enough to cover this change.

Signed-off-by: Gris Ge <fge@redhat.com>

tyll commented:
> @tyll Will it easier for you if I change `--skip-merge` to `--override`?

Override sounds better. After more thinking it also seems to me that this should be part of the State/YAML instead of a CLI option, for example by introducing something like

```
overriding_interfaces: 
- ...
```

that contains the states for the interfaces to override. Maybe there could be a better name for method for this like `interfaces@replace` so that the `@replace` marker could also be used in other settings to disable merging and we could introduce something like `@first` or `@last` that indicates that something would be added at a certain place, for example IP addresses or nameservers.

tyll commented:
The reason to put it into the state is because it is a property of the state (it is more a full state than a partial state) and this makes it immediately available to the user through all APIs.

cathay4t commented:
OpenShift installer use case require no changes to existing YAML files, hence changing schema will not solve the original problem.

cathay4t commented:
New patch uploaded. The option changed to `--override-iface` and will only override interfaces section now. For the route topic, I need to talk with customer to find better approaches.

ihuguet commented on doc/nmstate.service.8.in:
```diff
@@ -9,20 +9,43 @@ nmstate\&.service
 network stata to file with the suffix \fB.applied\fR after applying it.
 
+.PP
+By default, nmstate merge network state file with current network state,
+if you want to override interface settings without merging, please set
+`override_iface = true` in `[service]` section in
```
It is not clear what happens if override is used, please add a bried explanation like "default values will be used".

ihuguet commented on doc/nmstate.service.8.in:
```diff
@@ -9,20 +9,43 @@ nmstate\&.service
+\fR
+.EE
+
+Once `override_iface = true` been set, for network state to apply,
+nmstate.service will:
```
This sentence sounds weird to me... Maybe "If `override_iface = true`, for network state apply, nmstate.service will:"

ihuguet commented on doc/nmstate.service.8.in:
```diff
@@ -9,20 +9,43 @@ nmstate\&.service
+
+Once `override_iface = true` been set, for network state to apply,
+nmstate.service will:
+ * For `interfaces` section, any defined interface will be configured by desire
+   state information only.
```
Same than above: what happens with fields missing in the desired state?
Also, typo: desire -> desired

cathay4t replied:
Unified to:

```
 * For `interfaces` section, any defined interface will be configured by
   desired state information only, undefined properties of this interface will
   use default value instead of merging from current state.

 * For other sections, no changes to their original behavior.

For example, if desired interface has no IPv4 section defined, nmstate
will treat it as disabled regardless current network status.
```

ihuguet commented on doc/nmstatectl.8.in:
```diff
@@ -166,6 +166,14 @@ rebooting.
 the user must commit the changes within \fItimeout\fR, or they will be
 automatically rolled back. Default: 60 seconds.
+.IP \fB--override-iface\fR
+If defined, nmstate will not merge interface current state when applying
+desired state:
```
IMO, it would be clearer: nmstate will not merge the current network state to the interfaces' desired state.

cathay4t replied:
Changed to:

```
If defined when applying network state:

 * For `interfaces` section, any defined interface will be configured by
   desired state information only, undefined properties of this interface will
   use default value instead of merging from current state.
```

ihuguet commented on doc/nmstatectl.8.in:
```diff
@@ -166,6 +166,14 @@ rebooting.
+If defined, nmstate will not merge interface current state when applying
+desired state:
+
+ * For `interfaces` section, any defined interface will be configured by desire
+   state information only.
```
Same than above

cathay4t replied:
Changed to 
```
 * For `interfaces` section, any defined interface will be configured by
   desired state information only, undefined properties of this interface will
   use default value instead of merging from current state.
```

ihuguet commented on rust/src/lib/net_state.rs:
```diff
@@ -124,6 +124,8 @@ pub struct NetworkState {
     pub(crate) running_config_only: bool,
     #[serde(skip)]
     pub(crate) memory_only: bool,
+    #[serde(skip)]
+    pub(crate) override_iface: bool,
```
This is not super important, but I find a bit weird to set this in the NetworkState struct. It represents the current state of the network, so this option has no meaning there. It looks like more correct for MergedNetworkState, and having an argument `override_iface` in `NetworkState::apply`, where it does make sense.

As I say, this is not a must, though.

cathay4t replied:
Currently, the `apply()` function does not accept `NetworkStateApplyOption` due to my bad design, so user have to store the option via `NetworkState.set_xxx()`.

Will consider this is we are doing V2 for this.

ihuguet reviewed (commented):
Apart from those nits, LGTM.

ihuguet commented on doc/nmstatectl.8.in:
```diff
@@ -166,6 +166,18 @@ rebooting.
+.IP \fB--override-iface\fR
+If defined when applying netwrok state:
+
+ * For `interfaces` section, any defined interface will be configured by desire
```
desire -> desired. Apart from that, LGTM.

cathay4t replied:
Fixed. Thanks.

cathay4t committed:
apply: Support overriding interface configuration without merging

When using `nmstatectl gc`, unmentioned IP section means disabled, but
in `nmstatectl service` or `nmstatectl apply`, unmentioned IP means
merging from current. In order to be consistent when migrating from
`nmstatectl gc` to `nmstatectl service`, we need to introduce a option
to skip merging IP.

We also have user request us to not merging existing route by only
apply route found in desired state.

So we introduce option to skip merging current state:

 * Rust: `NetworkState.set_override_iface(true)`
 * Python: `libnmstate.apply(state, override_iface=True)`
 * C: `nmstate_net_state_apply()` with `NMSTATE_FLAG_OVERRIDE_IFACE` flag.
 * CLI: `nmstatectl apply --override-iface <state_file>`
 * nmstate.service: `/etc/nmstate/nmstate.conf` with:

    ```toml
    [service]
    override_iface = true
    ```

When in `override_iface` mode, nmstate will:

 * For `interfaces` section, any defined interface will be configured by
   desire state information only.
 * Other sections will ignore this `override_iface` mode and still
   merging as it was.

Technical detail:
 * When `override_iface: true`, we purge current network to only keep
   interface name and type before merging it with desire state.
 * When `override_iface: true`, we do not re-use existing NetworkManager
   connection.

Integration test cases included.

Resolves: https://issues.redhat.com/browse/RHEL-59935

Signed-off-by: Gris Ge <fge@redhat.com>

ffmancera reviewed (approved):
The PR is in quite good conditions. LGTM!

liangwen12year reviewed (approved):
I really like the highlight of technical details in the commit message, which makes me easier to understand the code changes.
